### DISCUSSION TITLE
Livewire Reusable Component

### DISCUSSION THREAD

**Discussion by ph7jack** [livewire/livewire#3380] [3. Feature ideas]
I'm creating a **POC** to pass a **wire:model** to a livewire component, and communicate with the parent component.
The idea is to be able to use **wire:model** the same way you use it on a blade component.

@calebporzio, what do you think?

Example Video
https://www.loom.com/share/b97c8ee04f2747c081f017b3c856ff98

**Comment by Dimonka2** [comment]
Is there a use case where you prefer to have it this way instead of "normal" blade component?

The only real use case I see is to have an advanced dropdown (with search, columns and sorting) in a separate component with integration into parent component. This way you can have dropdown logic done once and be reused in other components.

But you can do this integration via events.

**Reply by ph7jack** [reply]
This input is a simple example, but the real use case is a select component, with lazy loading data with search. 
A prop is a model class, value key, and label key. 

`<livewire:select wire:model="customerId" :model="\App\Models\Customer::class" option-label="name" option-value="id" />`

**Comment by joshhanley** [comment]
@PH7-Jack I'd love to see the implementation, do you have a branch on Github we can look at?

**Reply by ph7jack** [reply]
Yes, I will go make a commit. Wait a moment

**Reply by ph7jack** [reply]
@joshhanley https://github.com/livewire/livewire/compare/master...PH7-Jack:poc-wire-model?expand=1

is only a poc, a poor code to test this possibility

**Reply by ph7jack** [reply]
my testing component https://gist.github.com/PH7-Jack/49365e21b97a469b4ad88e0cf442d5d8

**Reply by joshhanley** [reply]
@PH7-Jack thanks! I will try and have a play over the weekend. It's a concept I've been thinking about for a while 🙂

If it can be implemented cleanly, it unlocks so much power with nested components and not having to use events all the time.

The one piece that would probably be needed though is an overhaul of how the request mechanism currently works. So multiple components can all request updates from the backend, but bundle it into a single request. That way we don't have to contend with race conditions of a parent or child request finishing first, which could potentially trigger a change and fire another request on the linked components.

It was what I aimed to enable through Sprucewire https://github.com/joshhanley/sprucewire
But that also had the same race conditions issue.

**Reply by ph7jack** [reply]
@joshhanley sprucewire is amazing!! 
I would really love to have this functionality as the default within livewire. I believe it is very useful for everyone.

**Comment by calebporzio** [comment]
Very interesting - this is something we've talked about extensively and I always come away feeling like it's an anti-pattern for Livewire and the best way is to use Blade components for these cases and wire:model to them 

This sort of thing will encourage more people to extract atomic Livewire components, which feels good in the moment, but requires extra roundtrips and makes the experience much more complex and slow.

I haven't looked at your strategy or implementation, but there is a possibility to allow this syntax and "couple" two components together so you aren't sending multiple network requests - but that would require some serious heavy technical lifting.

Anyhow, not to discourage you, it's totally possible you land on something really valuable here - just wanted to weigh in
